### PR TITLE
fix: low entropy mac passphrase in cipher seed (see issue #4182) 

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Description
---The following is an attempt to improve `CipherSeed` mnemonic generation by tackling MAC being keyed directly with a low entropy passphrase. We use proper domain separation to attain this.

Motivation and Context
--- The generation of MAC, within the context of a `CipherSeed` instance, is obtained through keying a (low) entropy passphrase. In order to reduce the chances of success of an attack involving offline key pre-computation, it is desirable to hash the passphrase, before MAC keying, using proper domain separation. The current PR is an attempt in this direction.

How Has This Been Tested?
--- With previous unit tests.

### CODE REVIEW

**Comment by jorgeantonio21** [comment]
Currently, we use a `Blake256` to generate domain separation. It might be desirable to use some other hashing algorithm, for purposes I don't understand.

Also feel free to comment below, as many of these concepts are not completely understood by the creator of the current PR.

File: applications/tari_validator_node/src/grpc/services/wallet_client.rs
Comment (SWvheerden) [inline applications/tari_validator_node/src/grpc/services/wallet_client.rs]: Why change this to a vec?
We are not changing this in the function?

File: dan_layer/core/src/services/checkpoint_manager.rs
Comment (SWvheerden) [inline dan_layer/core/src/services/checkpoint_manager.rs]: same here, why change to a vec?

File: dan_layer/core/src/services/checkpoint_manager.rs
Comment (SWvheerden) [inline dan_layer/core/src/services/checkpoint_manager.rs]: and here?

**Review by SWvheerden — COMMENTED** [review]
Why did you change a bunch of &[T] to Vec<T>?

**Comment by hansieodendaal** [comment]
Hi @jorgeantonio21, maybe you can squash all commits and rebase? There seems to be many unrelated changes in the code changes.

File: applications/tari_validator_node/src/grpc/services/wallet_client.rs
Comment (jorgeantonio21) [reply applications/tari_validator_node/src/grpc/services/wallet_client.rs]: I am sorry @SWvheerden, this was due to a merge conflict that I missed.

File: dan_layer/core/src/services/checkpoint_manager.rs
Comment (jorgeantonio21) [reply dan_layer/core/src/services/checkpoint_manager.rs]: Same merge conflict

File: dan_layer/core/src/services/checkpoint_manager.rs
Comment (jorgeantonio21) [reply dan_layer/core/src/services/checkpoint_manager.rs]: Same merge conflict

**Comment by jorgeantonio21** [comment]
> Hi @jorgeantonio21, maybe you can squash all commits and rebase? There seems to be many unrelated changes in the code changes.

Hi @hansieodendaal, thanks for the advice. I did rebase and the merge conflicts seem to be tackled.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (hansieodendaal) [inline base_layer/key_manager/src/cipher_seed.rs]: This limitation is not required anymore with the use of the hashing API

File: base_layer/key_manager/src/cipher_seed.rs
Comment (hansieodendaal) [inline base_layer/key_manager/src/cipher_seed.rs]:
```suggestion
const CIPHER_SEED_PERSONA: &str = "com.tari.key_manager.cipher_seed";
```

File: base_layer/key_manager/src/cipher_seed.rs
Comment (CjS77) [inline base_layer/key_manager/src/cipher_seed.rs]: Why not just use `Birthday` instead of tearing it apart?

File: base_layer/key_manager/src/cipher_seed.rs
Comment (CjS77) [inline base_layer/key_manager/src/cipher_seed.rs]: Instead of all this, just store a birthday instance

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: The way `version` is calculated seems very weird. Why is the version dependent on the current time??

There are no docs to explain this.

edit: Ok, it's just named in a very misleading way with no docs. See comments below

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: impl Default!

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]:
```suggestion
    fn new_from_time(network: Network, time: u64) -> Self {
```

if it was the current time you were using, you wouldn't need to pass the time in

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: It would. be more idiomatic to use `Duration` and `Instant`

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: ????

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: Is there a reason it's not `(days % PERIOD_LENGTH) as u16`?

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: I prefer `PERIOD_LENGTH * (version as u64)` but simply stylistic.

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [reply base_layer/key_manager/src/birthday.rs]: Ohhh! This is an epoch counter. 

Please rename it some other than version (epoch / period for example), and provide a doc string explaining what it is for.

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [inline base_layer/key_manager/src/birthday.rs]: I think it would be better to not have the dependency on `tari_core::blocks::genesis_block` at all. 
Better to pass in the genesis time as a parameter to the constructor. Then this module is more independent / decoupled. Then you ALSO can drop the dependence on `Network` it looks like. Bonus!

File: base_layer/key_manager/src/birthday.rs
Comment (CjS77) [reply base_layer/key_manager/src/birthday.rs]: And then wouldn't it be simpler to make the birthday a u32/u64 than manually reconstruct what is effectively a u24?

**Review by CjS77 — COMMENTED** [review]
Reviewed Birthday module primarily

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: That is definitely a possibility. However, I don't feel confident that the birthday genesis time might depend on the API use. Since we are primarily interested in using `Birthday` to set up a timestamp to scan the blockchain, it might be good to have a consistent genesis time for birthday. I am possibly missing something in your commentary.

Moreover, following [issue](https://github.com/tari-project/tari/issues/4176), it is referred that one should have an hard-coded birthday genesis time, and it is suggested to use the timestamp of the genesis network. I also really want to remove the dependency on the `Network`, as well. One could just hard-code a recent date (like now) and set it for the future use of the `Birthday` type.

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: Definitely so ! I am going to change it :)

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: Much more readable this way, I am going to change it.

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: I am going to add docs. When I implemented the code, I had in mind the third suggestion described [here](https://github.com/tari-project/tari/issues/4176). 

From what I understood, the `version` field in `Birthday` should really work as an epoch counter. This would be used later on to index the `version` in `CipherSeed`. Following this approach, versioning is epoch-dependent.

Do you think this approach makes sense  ?

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: Done

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: I added a more detailed comment below.

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: I agree. I only implemented this way for testing purposes. I am going to refactor it conveniently.

File: base_layer/key_manager/src/birthday.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/birthday.rs]: If one decides to keep the versioning for `CipherSeed` to be dependent on the epoch. Another approach would be to migrate the version computation to the `CipherSeed` impl, thus make it more explicit the use case we have in mind.

**Review by hansieodendaal — COMMENTED** [review]
LGTM, just the failing "test key manager wasm" tests left to fix

File: base_layer/key_manager/src/birthday.rs
Comment (AaronFeickert) [reply base_layer/key_manager/src/birthday.rs]: The intent of the overall `version` field should be an unambiguous identification of all parts of the protocol that all software agrees on. So if you were to change `Argon2` parameters or migrate to a different MAC algorithm, you'd define a new version; any software that knows about it will do the right thing, and any software that doesn't will simply produce an error.

I don't, however, think it's a good idea to attempt to compute a dynamic version number from the birthday. The intent of the comment in [this issue](https://github.com/tari-project/tari/issues/4176) was that once enough time passes such that a given birthday could potentially overflow, then you (or whoever is doing development that far into the future!) would manually define a new version that accounts for this. Doing it dynamically feels like a risky approach that could cause conflicts with any non-birthday protocol changes requiring new versions.

File: base_layer/key_manager/src/birthday.rs
Comment (AaronFeickert) [reply base_layer/key_manager/src/birthday.rs]: The idea behind hardcoding all parameters is to ensure that all implementations and installations agree on them unambiguously. Then you don't need to worry about differences in defaults or other computed parameters; everyone is working from the same playbook.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs:81]: To be clear, MAC-then-encrypt is in general not a safe approach to authenticated encryption. It can be shown to be safe under certain designs, like a stream cipher with a strong MAC, but it isn't clear that the truncated hash function approach here strictly meets this requirement.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: It's much more than just a hash of the password, which would not be safe for this purpose. The intent is to use a password-based key derivation function to generate a derived key of higher effective entropy through the use of a carefully-designed function like `Argon2` that's built for this purpose. Then we use separate portions of that derived key for the MAC and for `ChaCha20` encryption.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: For clarity I would rename this, since [HMAC](https://datatracker.ietf.org/doc/html/rfc2104) has a specific definition that isn't being used here. In this design the MAC is generated from a hash function, but it's not strictly HMAC.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: Is this domain separation label too generic? Other separators elsewhere are more detailed to avoid any accidental reuse elsewhere.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: A safer approach (and one used by the hashing API) is to ensure that any variable-length input to the underlying hash function be prepended with a fixed-length encoding of the input's length. This avoids accidental or malicious collisions. If you're taking a belt-and-suspenders approach, it's best to do this even for fixed-length inputs, in case they change later.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: Is there a more straightforward and/or more idiomatic way to do this?

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: Might be better to avoid the use of a magic number and move this length to a constant, in case it changes in a future version.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: Doesn't this take the _last_ 32 bytes?

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: As noted elsewhere, ideally this should prepend the length of the label using a fixed-length encoding.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: Thank you for the clarification here, Aaron. I reused part of your explanation in the documentation, I hope it makes more clear the goal of the current refactor.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: I rename it simply to `mac`.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: Definitely agree. I changed this to, following the current convention, `com.tari.key_manager.cipher_seed.chacha20_enc`. Similarly for the MAC generation label.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: Nice catch here :) !

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: I refactored it to use slicing, it makes the code cleaner and simpler to read. Thanks !

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: Yes, definitely a bug. Corrected !

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: I refactored the code to use the hashing API and take the first `n` byte elements, depending on the use case.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: I refactored the code to use the hashing API, this safety check should be taken care by the API.

File: base_layer/key_manager/src/cipher_seed.rs
Comment (hansieodendaal) [inline base_layer/key_manager/src/cipher_seed.rs:282]: Using the `MacDomain` from the hasing API:
```suggestion
    // TODO: Remove when generic `HashingDomain::mac_hasher` implementation is ready
    fn mac_hasher<D: Digest + LengthExtensionAttackResistant>(
        domain_label: &'static str,
    ) -> DomainSeparatedHasher<D, MacDomain> {
        DomainSeparatedHasher::<D, MacDomain>::new(domain_label)
    }

    fn generate_mac(
```

File: base_layer/key_manager/src/cipher_seed.rs
Comment (hansieodendaal) [inline base_layer/key_manager/src/cipher_seed.rs]: Using the `MacDomain` from the hasing API:
```suggestion
         // TODO: Replace with generic `HashingDomain::mac_hasher` implementation
        let mac = Self::mac_hasher::<Blake256>(DOMAIN_SEPARATION_LABEL)
```

**Review by hansieodendaal — COMMENTED** [review]
Hi, looking better and better. I have added a suggestion for using the `MacDomain` when generating the `mac`

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs:282]: Great suggestion !

File: base_layer/key_manager/src/cipher_seed.rs
Comment (AaronFeickert) [inline base_layer/key_manager/src/cipher_seed.rs]: This is missing the salt.

**Review by AaronFeickert — CHANGES_REQUESTED** [review]

File: base_layer/key_manager/src/cipher_seed.rs
Comment (jorgeantonio21) [reply base_layer/key_manager/src/cipher_seed.rs]: Totally my fault and a problematic bugs. Thanks for taking notice of it, I corrected accordingly!

**Review by hansieodendaal — DISMISSED** [review]
LGTM

File: base_layer/key_manager/src/lib.rs
Comment (hansieodendaal) [inline base_layer/key_manager/src/lib.rs]:
```suggestion
pub fn base_layer_key_manager() -> HashingDomain {
```

**Review by hansieodendaal — DISMISSED** [review]
LGTM - one naming conflict

File: base_layer/key_manager/src/lib.rs
Comment (stringhandler) [inline base_layer/key_manager/src/lib.rs]: I'd prefer something more like this:

```suggestion

hash_domain!(KeyManagerHashDomain, "com.tari.keymanager");

pub fn create_hasher<D:Digest>(label: &'static str) -> DomainSeparatedHasher<....> {
   DomainSeparatedHasher::new::<D>(label)
}




```

Comment (stringhandler) [reply base_layer/key_manager/src/lib.rs]: or using the `create_hasher!()` macro directly

Comment (stringhandler) [reply base_layer/key_manager/src/lib.rs]: or finally, one method for each use case...
```suggestion
hash_domain!(KeyManagerHashDomain, "com.tari.keymanager");

pub fn create_passphrase_hasher() -> DomainSeparatedHasher<VarBlake, KeyManagerHashDomain> {}

pub fn create_cipher_seed_mac_hasher() -> Domain....
```

**Review by stringhandler — APPROVED** [review]
Happy, but will wait for confirmation from @AaronFeickert  before merging

**Comment by AaronFeickert** [comment]
> Happy, but will wait for confirmation from @AaronFeickert before merging

A bit delayed (my apologies), but utACK.

**Pull request merged**
